File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n").filter(Boolean)

let index = 0

// z x y
const move = [
  [1, 0, 0],
  [-1, 0, 0],
  [0, 1, 0],
  [0, -1, 0],
  [0, 0, 1],
  [0, 0, -1],
]
// 상 하 북 남 동 서

const bfs = (graph,startPosition, L, R, C,endPosition) => {
  const queue = [...startPosition]
  const [tz,tx,ty] = endPosition[0]
  const visited = Array.from({ length: L }, () => Array.from({ length: R }, () => Array.from({ length: C }, () => -1)))
  for (let [z, x, y] of startPosition) {
    visited[z][x][y] = 0
  }

  let index = 0
  while (queue.length > index) {
    const [z, x, y] = queue[index++]
    for (let i = 0; i < 6; i++) {
      const nz = z + move[i][0]
      const nx = x + move[i][1]
      const ny = y + move[i][2]
      if (
        nx < 0 ||
        ny < 0 ||
        nz < 0 ||
        nz >= L ||
        nx >= R ||
        ny >= C ||
        graph[nz][nx][ny] === "#" ||
        visited[nz][nx][ny] !== -1
      ) {
        continue
      }
      visited[nz][nx][ny] = visited[z][x][y] + 1
      queue.push([nz, nx, ny])
    }
  }
  return visited[tz][tx][ty]
}
const answer = []
while (true) {
  const [L, R, C] = input[index++].split(" ").map(Number)
  if (L === 0 && R === 0 && C === 0) break
  const graph = []
  const startPosition = []
  const endPosition = []
  for (let i = 0; i < L; i++) {
    graph.push(input.slice(index, index + R).map((row) => row.split("")))
    index += R
  }
  for (let i = 0; i < L; i++) {
    for (let j = 0; j < R; j++) {
      for (let k = 0; k < C; k++) {
        if (graph[i][j][k] === "S") {
          startPosition.push([i, j, k])
        } else if (graph[i][j][k] === "E") {
          endPosition.push([i, j, k])
        }
      }
    }
  }
  const result = bfs(graph,startPosition,L,R,C,endPosition)
  if(result === -1){
    answer.push("Trapped!")
  }else{
    answer.push(`Escaped in ${result} minute(s).`)
  }
}

console.log(answer.join('\n'))

File: 백준/Gold/6593. 상범 빌딩/상범 빌딩.py
import sys
from collections import deque

input = sys.stdin.readline

# z, x, y 이동 방향
move = [
    (1, 0, 0), (-1, 0, 0),  # 상, 하
    (0, 1, 0), (0, -1, 0),  # 북, 남
    (0, 0, 1), (0, 0, -1)   # 동, 서
]

def bfs(graph, start_positions, L, R, C, end_position):
    queue = deque(start_positions)
    visited = [[[-1] * C for _ in range(R)] for _ in range(L)]
    
    # 시작 지점 방문 처리
    for z, x, y in start_positions:
        visited[z][x][y] = 0

    while queue:
        z, x, y = queue.popleft()
        for dz, dx, dy in move:
            nz, nx, ny = z + dz, x + dx, y + dy
            if 0 <= nz < L and 0 <= nx < R and 0 <= ny < C:
                if graph[nz][nx][ny] != "#" and visited[nz][nx][ny] == -1:
                    visited[nz][nx][ny] = visited[z][x][y] + 1
                    queue.append((nz, nx, ny))

    # 도착 지점의 거리 반환
    tz, tx, ty = end_position[0]
    return visited[tz][tx][ty]

answer = []

while True:
    # L, R, C 입력
    L, R, C = map(int, input().strip().split())
    if L == 0 and R == 0 and C == 0:
        break

    graph = []
    start_positions = []
    end_position = []

    # 3차원 그래프 입력 처리
    for i in range(L):
        layer = []
        for j in range(R):
            row = list(input().strip())
            layer.append(row)
        graph.append(layer)
        input()  # 빈 줄 건너뛰기

    # 시작점과 도착점 찾기
    for i in range(L):
        for j in range(R):
            for k in range(C):
                if graph[i][j][k] == "S":
                    start_positions.append((i, j, k))
                elif graph[i][j][k] == "E":
                    end_position.append((i, j, k))

    # BFS 실행
    result = bfs(graph, start_positions, L, R, C, end_position)
    if result == -1:
        answer.append("Trapped!")
    else:
        answer.append(f"Escaped in {result} minute(s).")

# 정답 출력
print("\n".join(answer))
